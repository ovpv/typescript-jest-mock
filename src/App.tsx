import React, { useState } from "react";
import "./styles.css";
import moment from "moment";

export default function App() {
  const [date, setDate] = useState("");
  const [age, setAge] = useState(0);
  const calculateAge = () => {
    console.log('diff called');
    const calcAge = moment().diff(date, "years");
    setAge(calcAge);
  };
  return (
    <div className="App">
      <h2>How old are you?</h2>
      <label>Birthday: </label>
      <input
        type="date"
        onChange={(e) => setDate(e.target.value)}
        value={date}
        id="birthday"
        name="birthday"
      />
      <button disabled={!date.length} onClick={calculateAge}>
        Calculate Age
      </button>
      <div>Your age is: {age}</div>
    </div>
  );
}
