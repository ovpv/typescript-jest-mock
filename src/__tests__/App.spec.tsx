import { shallow, configure } from "enzyme";
import React from "react";
import App from "../App";
import Adapter from 'enzyme-adapter-react-16';

//define mock for moment lib
const diff = jest.fn();
const momentObj = () => ({
  diff
})
jest.mock('moment', () => momentObj)

configure({ adapter: new Adapter() });

describe("App", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<App />);
  });
  it("should render properly", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('should have calculate button disabled, if date is not selected', () => {
    const calcButton = wrapper.find('button').at(0);
    expect(calcButton.prop('disabled')).toBe(true);
  });
  it('should have calculate button enabled after date is selected', () => {
    const picker = wrapper.find('input').at(0);
    picker.prop('onChange')({
      target: {
        value: '1990-10-02'
      }
    });
    const calcButton = wrapper.find('button').at(0);
    expect(calcButton.prop('disabled')).toBe(false);
  });
  it('should execute moment method to calculate age on button click', () => {
    const moment = require('moment');
    const picker = wrapper.find('input').at(0);
    picker.prop('onChange')({
      target: {
        value: '1990-10-02'
      }
    });
    const calcButton = wrapper.find('button').at(0);
    calcButton.prop('onClick')();
    expect(diff).toHaveBeenCalled();
  })
});
