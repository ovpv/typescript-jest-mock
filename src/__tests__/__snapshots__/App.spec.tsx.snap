// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render properly 1`] = `
<div
  className="App"
>
  <h2>
    How old are you?
  </h2>
  <label>
    Birthday: 
  </label>
  <input
    id="birthday"
    name="birthday"
    onChange={[Function]}
    type="date"
    value=""
  />
  <button
    disabled={true}
    onClick={[Function]}
  >
    Calculate Age
  </button>
  <div>
    Your age is: 
    0
  </div>
</div>
`;
